README.md
[![time tracker](https://wakatime.com/badge/github/SomeoneSom/MindusIdle.svg)](https://wakatime.com/badge/github/SomeoneSom/MindusIdle)
[![codebeat badge](https://codebeat.co/badges/0a6ba37b-ccc7-43c2-98b6-d98c8a171d8d)](https://codebeat.co/projects/github-com-someonesom-mindusidle-gh-pages)
# MindusIdle
Idle game based on Mindustry
Check it out [here](https://someonesom.github.io/MindusIdle)!


changelog.txt
v0.6.0 - 1/20/21
    Offline progress is now a thing
    Fixed bug where it would not autosave every 30 seconds

v0.5.3 - 1/19/21
    Added Copper Excavator
    Added Lead Excavator

v0.5.2 - 1/19/21
    Fixed bug where power didnt update

v0.5.1 - 1/19/21
    Buildings now quadruple in power every 25 purchases from 100-1000, and by ten every 100 purchases from 1000-10000

v0.5.0 - 1/12/21
	Bugs with power have been fixed
	Added Combustion Generator

v0.4.0 - 1/8/21
    Added power (WARNING: NOT TESTED)
    Removed the word "shop" from navbar

v0.3.0 - 1/8/21
    Added jQuery Style Switcher
    Added (ugly) light theme

v0.2.1 - 1/7/21
    Fixed major bug where crafting shop text persisted on save wipe

v0.2.0 - 1/7/21
    Added Coal Drill
    Added Graphite Press
    Changed styling of buttons slightly

v0.1.2 - 1/7/21
    Version text is now in bottom right as not not be covered by hovering over anchors
    Drills now unlock when you meet the requirements to buy them

v0.1.1 - 1/6/21
    Fixed major bug that overwrote progress
    
v0.1.0 - 1/6/21
    Made changelog.txt
    Fixed making new save data if there was none
    Now gets new features automatically instead of having to wipe save
    Version, source code link, and changelog link now in bottom left

scripts/main.js
var divsmain = ["res", "drshop", "crshop", "upg", "power", "land", "options"];
var resources;
var obj;
function tick(times = 1) {
  for (var k = 0; k < times; k++) {
    for (var i = 0; i < Object.keys(obj).length; i++) {
      object = obj[Object.keys(obj)[i]]
      amount = object[0];
      for (var j = 0; j < 17; j++) {
        if (object[5]*amount <= resources["power"]) {
          resources[Object.keys(resources)[j]] += object[2][j] * amount;
          resources["power"] -= object[5]*amount;
        }
      }
    }
  }
}
function enableDiv(divname, divs) {
  for (var j = 0; j < divs.length; j++) {
    var divtag = divs[j];
    var div = document.getElementById(divtag);
    var anchor = document.getElementsByClassName(divtag)[0];
    if (divname == divtag) {
      div.style.display = "block";
      anchor.style.color = "orange";
    } else {
      div.style.display = "none";
      anchor.style.color = "white";
    }
  }
}
enableDiv("mech", divsdr);
enableDiv("res", divsmain);
enableDiv("duo", divslnd);
function wipeSave() {
  window.localStorage['resources'] = resourcesj;
  window.localStorage['obj'] = objj;
  window.localStorage['upgrades'] = upgradesj;
  document.getElementById('upg').innerHTML = "";
}
function loadSave() {
  resources = JSON.parse(window.localStorage['resources']);
  obj = JSON.parse(window.localStorage['obj']);
  upgrades = JSON.parse(window.localStorage['upgrades']);
}
function getNewFeatures() {
  for (var objct of Object.keys(JSON.parse(objj))) {
    if (typeof(obj[objct]) == "undefined") {
      obj[objct] = JSON.parse(objj)[objct];
    }
    for (var i = 1; i < 6; i++) {
      obj[objct][i] = JSON.parse(objj)[objct][i];
    }
  }
  for (var resn of Object.keys(JSON.parse(resourcesj))) {
    if (typeof(resources[resn]) == "undefined") {
      resources[resn] = JSON.parse(resourcesj)[resn];
    }
  }
}
if (typeof(window.localStorage['resources']) == "undefined") {
  wipeSave();
}
function save() {
  window.localStorage['resources'] = JSON.stringify(resources);
  window.localStorage['obj'] = JSON.stringify(obj);
  //this code shows the 'Saved!' bar
  var bar = document.getElementsByClassName("autosavebar")[0];
  bar.style.visibility = "visible";
  bar.style.opacity = 1;
  window.setTimeout(function(){bar.style.visibility = "hidden";bar.style.opacity = 0;}, 1000);
}
function buy(objc, amount) {
  for (var i = 0; i < 17; i++) {
    if (resources[Object.keys(resources)[i]] < objc[1][i]) {
      return;
    }
  }
  for (var i = 0; i < 17; i++) {
    resources[Object.keys(resources)[i]] -= objc[1][i];
    if (i != 16) {
      objc[1][i] = Math.round(objc[1][i] * 1.15);
    }
  }
  objc[0] += 1;
  if (objc[0] % 100 == 0 && objc[0] >= 1000 && objc[0] <= 10000) {
    objc[2] = objc[2].map(x => x*10);
  } else if (objc[0] % 25 == 0 && objc[0] >= 100 && objc[0] < 1000) {
    objc[2] = objc[2].map(x => x*4);
  }
}
loadSave();
getNewFeatures();
if (!localStorage['lastAccessed']) {
  localStorage['lastAccessed'] = Math.floor(Date.now() / 1000);
}
var disparity = Math.floor(Date.now() / 1000) - localStorage['lastAccessed'];
localStorage['lastAccessed'] = Math.floor(Date.now() / 1000);
tick(disparity);
//nums go up here
window.setInterval(tick, 1000);
//building and res counts change here
window.setInterval(function(){
  for (var i = 0; i < 18; i++) {
    document.getElementById(i.toString(10).concat("res")).innerHTML = parseFloat(resources[Object.keys(resources)[i]].toFixed(2)).toString(10);
  }
  for (var i = 0; i < Object.keys(obj).length; i++) {
    document.getElementById(i.toString(10).concat("count")).innerHTML = obj[Object.keys(obj)[i]][0].toString(10);
    for (var j = 0; j < 17; j++) {
      out = obj[Object.keys(obj)[i]][2][j];
      cost = obj[Object.keys(obj)[i]][1][j].toString(10);
      if (cost != "0") {
        document.getElementById(i.toString(10).concat("cost", j.toString(10))).innerHTML = cost;
      }
      if (out != 0) {
        out *= obj[Object.keys(obj)[i]][0];
        document.getElementById(i.toString(10).concat("out", j.toString(10))).innerHTML = out.toFixed(1); 
      }
    }
  }
}, 100);
//needed for upgrades
function do_func(func) {
  func();
}
//autosave every 30 seconds
window.setInterval(save, 30000);
//upgrade management
window.setInterval(function(){
  for (var i = 0; i < upgrades.length; i++) {
    if ((eval('('+upgrades[i][3]+'());') == true) && (document.getElementById(upgrades[i][0]) == null) && (upgrades)) {
      document.getElementById('upg').innerHTML += `<button id="${upgrades[i][0]}" title="${upgrades[i][1]}" onclick="do_func(${upgrades[i][4]});upgrades[${i}][2]=1;">${upgrades[i][0]}</button>`;
      upgrades[i][5] = 1;    
    }
    if (upgrades[i][2] == 1 && document.getElementById(upgrades[i][0]) != null) {
      element = document.getElementById(upgrades[i][0]);
      element.parentNode.removeChild(element);
    }
  }
}, 100);
//save on window close
window.addEventListener('beforeunload', function(){
  save();
  return null;
});

scripts/template.js
//capitalize function for obj names
String.prototype.capitalize = function() {
    return this.charAt(0).toUpperCase() + this.slice(1);
};
//make all necessary anchors and divs in the correct locations for all the objs
var divsdr = divscr = divslnd = divspower = [];
var i = 0;
for (var objct of Object.keys(JSON.parse(objj))) {
    var flag = JSON.parse(objj)[objct][3];
    if (flag == 0) {
        var divlst = "divsdr";
        var id = "drnav";
        document.getElementById('drshop').innerHTML += `<div id="${objct}"></div>`;
        divsdr.push(objct);
    } else if (flag == 1) {
        var divlst = "divscr";
        var id = "crnav";
        document.getElementById('crshop').innerHTML += `<div id="${objct}"></div>`;
        divscr.push(objct);
    } else if (flag == 2) {
        var divlst = "divslnd";
        var id = "lndnav"
        document.getElementById('land').innerHTML += `<div id="${objct}"></div>`;
        divslnd.push(objct);
    } else {
        var divlst = "divspower";
        var id = "powernav";
        document.getElementById('power').innerHTML += `<div id="${objct}"></div>`;
        divspower.push(objct);
    }
    document.getElementById(id).innerHTML += `<li><a href="javascript:enableDiv('${objct}', ${divlst})" class="${objct}">${JSON.parse(objj)[objct][4]}</a></li>`;
    document.getElementById(objct).innerHTML += `<p>You have <span id="${i}count">0</span> ${pluralize(JSON.parse(objj)[objct][4])} ${(flag != 3) ? "producing" : "."}:`;
    for (var j = 0; j < 17; j++) {
        if (JSON.parse(objj)[objct][2][j] != 0) {
            document.getElementById(objct).innerHTML += `<span id="${i}out${j}"> 0</span> ${Object.keys(JSON.parse(resourcesj))[j].capitalize()} `;
        }
    }
    document.getElementById(objct).innerHTML += `${(flag != 3) ? 'per second.' : ''}<br /><br />
    Buying 1 will cost:<br /><br />`;
    for (var j = 0; j < 17; j++) {
        if (JSON.parse(objj)[objct][1][j] != 0) {
            document.getElementById(objct).innerHTML += `<span id="${i}cost${j}"> 0</span> ${Object.keys(JSON.parse(resourcesj))[j].capitalize()} `;
        }
    }
    document.getElementById(objct).innerHTML += `<br /><br />
    Additionally, each ${JSON.parse(objj)[objct][4]} ${(JSON.parse(objj)[objct][5] < 0) ? 'produces' : 'consumes'} ${Math.abs(JSON.parse(objj)[objct][5])} power per second, totaling <span id="${i}out{j}">0</span> power per second.</p>
    <button onclick="buy(obj.${objct}, 1);">Buy 1</button>`;
    i++;
}
//manage unlocking of all objects (not upgrades)
window.setInterval(function(){
    for (var objct of Object.keys(JSON.parse(objj))) {
        if (obj[objct][0] > 0) {
            continue;
        }
        var entry = document.getElementsByClassName(objct)[0];
        var canBuy = JSON.parse(objj)[objct][1].map(function(x, i){return x <= Object.values(resources)[i]}).every(Boolean);
        if (canBuy == true) {
            entry.style.display = "";
        } else {
            entry.style.display = "none";
            document.getElementById(objct).style.display = "none";
        }
    }
}, 100);

modules/pluralize.js
/* global define */

(function (root, pluralize) {
    /* istanbul ignore else */
    if (typeof require === 'function' && typeof exports === 'object' && typeof module === 'object') {
      // Node.
      module.exports = pluralize();
    } else if (typeof define === 'function' && define.amd) {
      // AMD, registers as an anonymous module.
      define(function () {
        return pluralize();
      });
    } else {
      // Browser global.
      root.pluralize = pluralize();
    }
  })(this, function () {
    // Rule storage - pluralize and singularize need to be run sequentially,
    // while other rules can be optimized using an object for instant lookups.
    var pluralRules = [];
    var singularRules = [];
    var uncountables = {};
    var irregularPlurals = {};
    var irregularSingles = {};
  
    /**
     * Sanitize a pluralization rule to a usable regular expression.
     *
     * @param  {(RegExp|string)} rule
     * @return {RegExp}
     */
    function sanitizeRule (rule) {
      if (typeof rule === 'string') {
        return new RegExp('^' + rule + '$', 'i');
      }
  
      return rule;
    }
  
    /**
     * Pass in a word token to produce a function that can replicate the case on
     * another word.
     *
     * @param  {string}   word
     * @param  {string}   token
     * @return {Function}
     */
    function restoreCase (word, token) {
      // Tokens are an exact match.
      if (word === token) return token;
  
      // Lower cased words. E.g. "hello".
      if (word === word.toLowerCase()) return token.toLowerCase();
  
      // Upper cased words. E.g. "WHISKY".
      if (word === word.toUpperCase()) return token.toUpperCase();
  
      // Title cased words. E.g. "Title".
      if (word[0] === word[0].toUpperCase()) {
        return token.charAt(0).toUpperCase() + token.substr(1).toLowerCase();
      }
  
      // Lower cased words. E.g. "test".
      return token.toLowerCase();
    }
  
    /**
     * Interpolate a regexp string.
     *
     * @param  {string} str
     * @param  {Array}  args
     * @return {string}
     */
    function interpolate (str, args) {
      return str.replace(/\$(\d{1,2})/g, function (match, index) {
        return args[index] || '';
      });
    }
  
    /**
     * Replace a word using a rule.
     *
     * @param  {string} word
     * @param  {Array}  rule
     * @return {string}
     */
    function replace (word, rule) {
      return word.replace(rule[0], function (match, index) {
        var result = interpolate(rule[1], arguments);
  
        if (match === '') {
          return restoreCase(word[index - 1], result);
        }
  
        return restoreCase(match, result);
      });
    }
  
    /**
     * Sanitize a word by passing in the word and sanitization rules.
     *
     * @param  {string}   token
     * @param  {string}   word
     * @param  {Array}    rules
     * @return {string}
     */
    function sanitizeWord (token, word, rules) {
      // Empty string or doesn't need fixing.
      if (!token.length || uncountables.hasOwnProperty(token)) {
        return word;
      }
  
      var len = rules.length;
  
      // Iterate over the sanitization rules and use the first one to match.
      while (len--) {
        var rule = rules[len];
  
        if (rule[0].test(word)) return replace(word, rule);
      }
  
      return word;
    }
  
    /**
     * Replace a word with the updated word.
     *
     * @param  {Object}   replaceMap
     * @param  {Object}   keepMap
     * @param  {Array}    rules
     * @return {Function}
     */
    function replaceWord (replaceMap, keepMap, rules) {
      return function (word) {
        // Get the correct token and case restoration functions.
        var token = word.toLowerCase();
  
        // Check against the keep object map.
        if (keepMap.hasOwnProperty(token)) {
          return restoreCase(word, token);
        }
  
        // Check against the replacement map for a direct word replacement.
        if (replaceMap.hasOwnProperty(token)) {
          return restoreCase(word, replaceMap[token]);
        }
  
        // Run all the rules against the word.
        return sanitizeWord(token, word, rules);
      };
    }
  
    /**
     * Check if a word is part of the map.
     */
    function checkWord (replaceMap, keepMap, rules, bool) {
      return function (word) {
        var token = word.toLowerCase();
  
        if (keepMap.hasOwnProperty(token)) return true;
        if (replaceMap.hasOwnProperty(token)) return false;
  
        return sanitizeWord(token, token, rules) === token;
      };
    }
  
    /**
     * Pluralize or singularize a word based on the passed in count.
     *
     * @param  {string}  word      The word to pluralize
     * @param  {number}  count     How many of the word exist
     * @param  {boolean} inclusive Whether to prefix with the number (e.g. 3 ducks)
     * @return {string}
     */
    function pluralize (word, count, inclusive) {
      var pluralized = count === 1
        ? pluralize.singular(word) : pluralize.plural(word);
  
      return (inclusive ? count + ' ' : '') + pluralized;
    }
  
    /**
     * Pluralize a word.
     *
     * @type {Function}
     */
    pluralize.plural = replaceWord(
      irregularSingles, irregularPlurals, pluralRules
    );
  
    /**
     * Check if a word is plural.
     *
     * @type {Function}
     */
    pluralize.isPlural = checkWord(
      irregularSingles, irregularPlurals, pluralRules
    );
  
    /**
     * Singularize a word.
     *
     * @type {Function}
     */
    pluralize.singular = replaceWord(
      irregularPlurals, irregularSingles, singularRules
    );
  
    /**
     * Check if a word is singular.
     *
     * @type {Function}
     */
    pluralize.isSingular = checkWord(
      irregularPlurals, irregularSingles, singularRules
    );
  
    /**
     * Add a pluralization rule to the collection.
     *
     * @param {(string|RegExp)} rule
     * @param {string}          replacement
     */
    pluralize.addPluralRule = function (rule, replacement) {
      pluralRules.push([sanitizeRule(rule), replacement]);
    };
  
    /**
     * Add a singularization rule to the collection.
     *
     * @param {(string|RegExp)} rule
     * @param {string}          replacement
     */
    pluralize.addSingularRule = function (rule, replacement) {
      singularRules.push([sanitizeRule(rule), replacement]);
    };
  
    /**
     * Add an uncountable word rule.
     *
     * @param {(string|RegExp)} word
     */
    pluralize.addUncountableRule = function (word) {
      if (typeof word === 'string') {
        uncountables[word.toLowerCase()] = true;
        return;
      }
  
      // Set singular and plural references for the word.
      pluralize.addPluralRule(word, '$0');
      pluralize.addSingularRule(word, '$0');
    };
  
    /**
     * Add an irregular word definition.
     *
     * @param {string} single
     * @param {string} plural
     */
    pluralize.addIrregularRule = function (single, plural) {
      plural = plural.toLowerCase();
      single = single.toLowerCase();
  
      irregularSingles[single] = plural;
      irregularPlurals[plural] = single;
    };
  
    /**
     * Irregular rules.
     */
    [
      // Pronouns.
      ['I', 'we'],
      ['me', 'us'],
      ['he', 'they'],
      ['she', 'they'],
      ['them', 'them'],
      ['myself', 'ourselves'],
      ['yourself', 'yourselves'],
      ['itself', 'themselves'],
      ['herself', 'themselves'],
      ['himself', 'themselves'],
      ['themself', 'themselves'],
      ['is', 'are'],
      ['was', 'were'],
      ['has', 'have'],
      ['this', 'these'],
      ['that', 'those'],
      // Words ending in with a consonant and `o`.
      ['echo', 'echoes'],
      ['dingo', 'dingoes'],
      ['volcano', 'volcanoes'],
      ['tornado', 'tornadoes'],
      ['torpedo', 'torpedoes'],
      // Ends with `us`.
      ['genus', 'genera'],
      ['viscus', 'viscera'],
      // Ends with `ma`.
      ['stigma', 'stigmata'],
      ['stoma', 'stomata'],
      ['dogma', 'dogmata'],
      ['lemma', 'lemmata'],
      ['schema', 'schemata'],
      ['anathema', 'anathemata'],
      // Other irregular rules.
      ['ox', 'oxen'],
      ['axe', 'axes'],
      ['die', 'dice'],
      ['yes', 'yeses'],
      ['foot', 'feet'],
      ['eave', 'eaves'],
      ['goose', 'geese'],
      ['tooth', 'teeth'],
      ['quiz', 'quizzes'],
      ['human', 'humans'],
      ['proof', 'proofs'],
      ['carve', 'carves'],
      ['valve', 'valves'],
      ['looey', 'looies'],
      ['thief', 'thieves'],
      ['groove', 'grooves'],
      ['pickaxe', 'pickaxes'],
      ['passerby', 'passersby']
    ].forEach(function (rule) {
      return pluralize.addIrregularRule(rule[0], rule[1]);
    });
  
    /**
     * Pluralization rules.
     */
    [
      [/s?$/i, 's'],
      [/[^\u0000-\u007F]$/i, '$0'],
      [/([^aeiou]ese)$/i, '$1'],
      [/(ax|test)is$/i, '$1es'],
      [/(alias|[^aou]us|t[lm]as|gas|ris)$/i, '$1es'],
      [/(e[mn]u)s?$/i, '$1s'],
      [/([^l]ias|[aeiou]las|[ejzr]as|[iu]am)$/i, '$1'],
      [/(alumn|syllab|vir|radi|nucle|fung|cact|stimul|termin|bacill|foc|uter|loc|strat)(?:us|i)$/i, '$1i'],
      [/(alumn|alg|vertebr)(?:a|ae)$/i, '$1ae'],
      [/(seraph|cherub)(?:im)?$/i, '$1im'],
      [/(her|at|gr)o$/i, '$1oes'],
      [/(agend|addend|millenni|dat|extrem|bacteri|desiderat|strat|candelabr|errat|ov|symposi|curricul|automat|quor)(?:a|um)$/i, '$1a'],
      [/(apheli|hyperbat|periheli|asyndet|noumen|phenomen|criteri|organ|prolegomen|hedr|automat)(?:a|on)$/i, '$1a'],
      [/sis$/i, 'ses'],
      [/(?:(kni|wi|li)fe|(ar|l|ea|eo|oa|hoo)f)$/i, '$1$2ves'],
      [/([^aeiouy]|qu)y$/i, '$1ies'],
      [/([^ch][ieo][ln])ey$/i, '$1ies'],
      [/(x|ch|ss|sh|zz)$/i, '$1es'],
      [/(matr|cod|mur|sil|vert|ind|append)(?:ix|ex)$/i, '$1ices'],
      [/\b((?:tit)?m|l)(?:ice|ouse)$/i, '$1ice'],
      [/(pe)(?:rson|ople)$/i, '$1ople'],
      [/(child)(?:ren)?$/i, '$1ren'],
      [/eaux$/i, '$0'],
      [/m[ae]n$/i, 'men'],
      ['thou', 'you']
    ].forEach(function (rule) {
      return pluralize.addPluralRule(rule[0], rule[1]);
    });
  
    /**
     * Singularization rules.
     */
    [
      [/s$/i, ''],
      [/(ss)$/i, '$1'],
      [/(wi|kni|(?:after|half|high|low|mid|non|night|[^\w]|^)li)ves$/i, '$1fe'],
      [/(ar|(?:wo|[ae])l|[eo][ao])ves$/i, '$1f'],
      [/ies$/i, 'y'],
      [/(dg|ss|ois|lk|ok|wn|mb|th|ch|ec|oal|is|ck|ix|sser|ts|wb)ies$/i, '$1ie'],
      [/\b(l|(?:neck|cross|hog|aun)?t|coll|faer|food|gen|goon|group|hipp|junk|vegg|(?:pork)?p|charl|calor|cut)ies$/i, '$1ie'],
      [/\b(mon|smil)ies$/i, '$1ey'],
      [/\b((?:tit)?m|l)ice$/i, '$1ouse'],
      [/(seraph|cherub)im$/i, '$1'],
      [/(x|ch|ss|sh|zz|tto|go|cho|alias|[^aou]us|t[lm]as|gas|(?:her|at|gr)o|[aeiou]ris)(?:es)?$/i, '$1'],
      [/(analy|diagno|parenthe|progno|synop|the|empha|cri|ne)(?:sis|ses)$/i, '$1sis'],
      [/(movie|twelve|abuse|e[mn]u)s$/i, '$1'],
      [/(test)(?:is|es)$/i, '$1is'],
      [/(alumn|syllab|vir|radi|nucle|fung|cact|stimul|termin|bacill|foc|uter|loc|strat)(?:us|i)$/i, '$1us'],
      [/(agend|addend|millenni|dat|extrem|bacteri|desiderat|strat|candelabr|errat|ov|symposi|curricul|quor)a$/i, '$1um'],
      [/(apheli|hyperbat|periheli|asyndet|noumen|phenomen|criteri|organ|prolegomen|hedr|automat)a$/i, '$1on'],
      [/(alumn|alg|vertebr)ae$/i, '$1a'],
      [/(cod|mur|sil|vert|ind)ices$/i, '$1ex'],
      [/(matr|append)ices$/i, '$1ix'],
      [/(pe)(rson|ople)$/i, '$1rson'],
      [/(child)ren$/i, '$1'],
      [/(eau)x?$/i, '$1'],
      [/men$/i, 'man']
    ].forEach(function (rule) {
      return pluralize.addSingularRule(rule[0], rule[1]);
    });
  
    /**
     * Uncountable rules.
     */
    [
      // Singular words with no plurals.
      'adulthood',
      'advice',
      'agenda',
      'aid',
      'aircraft',
      'alcohol',
      'ammo',
      'analytics',
      'anime',
      'athletics',
      'audio',
      'bison',
      'blood',
      'bream',
      'buffalo',
      'butter',
      'carp',
      'cash',
      'chassis',
      'chess',
      'clothing',
      'cod',
      'commerce',
      'cooperation',
      'corps',
      'debris',
      'diabetes',
      'digestion',
      'elk',
      'energy',
      'equipment',
      'excretion',
      'expertise',
      'firmware',
      'flounder',
      'fun',
      'gallows',
      'garbage',
      'graffiti',
      'hardware',
      'headquarters',
      'health',
      'herpes',
      'highjinks',
      'homework',
      'housework',
      'information',
      'jeans',
      'justice',
      'kudos',
      'labour',
      'literature',
      'machinery',
      'mackerel',
      'mail',
      'media',
      'mews',
      'moose',
      'music',
      'mud',
      'manga',
      'news',
      'only',
      'personnel',
      'pike',
      'plankton',
      'pliers',
      'police',
      'pollution',
      'premises',
      'rain',
      'research',
      'rice',
      'salmon',
      'scissors',
      'series',
      'sewage',
      'shambles',
      'shrimp',
      'software',
      'staff',
      'swine',
      'tennis',
      'traffic',
      'transportation',
      'trout',
      'tuna',
      'wealth',
      'welfare',
      'whiting',
      'wildebeest',
      'wildlife',
      'you',
      /pok[eé]mon$/i,
      // Regexes.
      /[^aeiou]ese$/i, // "chinese", "japanese"
      /deer$/i, // "deer", "reindeer"
      /fish$/i, // "fish", "blowfish", "angelfish"
      /measles$/i,
      /o[iu]s$/i, // "carnivorous"
      /pox$/i, // "chickpox", "smallpox"
      /sheep$/i
    ].forEach(pluralize.addUncountableRule);
  
    return pluralize;
  });

scripts/data.js
var resourcesj = JSON.stringify({
    "copper":25,
    "lead":0,
    "scrap":0,
    "sand":0,
    "graphite":0,
    "metaglass":0,
    "spore_pod":0,
    "coal":0,
    "titanium":0,
    "thorium":0,
    "silicon":0,
    "plastanium":0,
    "phase_fabric":0,
    "surge_alloy":0,
    "blast_compound":0,
    "pyratite":0,
    "land":10,
    "power":0
});
var objj = JSON.stringify({
    "mech": [
        0,
        [25, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
        [1, 0.1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        0,
        "Mechanical Drill",
        0
    ],
    "duo": [
        0,
        [35, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0.1],
        2,
        "Duo",
        0
    ],
    "leadr": [
        0,
        [2, 30, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
        [0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        0,
        "Lead Drill",
        0
    ],
    "coaldr": [
        0,
        [100, 100, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 1.5, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        0,
        "Coal Drill",
        0
    ],
    "graphpress": [
        0,
        [150, 150, 0, 0, 0, 0, 0, 150, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        1,
        "Graphite Press",
        0
    ],
    "combusgen": [
        0,
        [0, 0, 0, 0, 100, 0, 0, 200, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        3,
        "Combustion Generator",
        -60
    ],
    "coppexca": [
        0,
        [250, 10, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [15, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        0,
        "Copper Excavator",
        30
    ],
    "leadexca": [
        0,
        [50, 350, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 15, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        0,
        "Lead Excavator",
        30
    ]
});
var upgradesj = JSON.stringify([
    [
        "Extra-Powerful Grease",
        "Speed up Mechanical Drills by 2x.\nCosts 100 Copper.",
        0,
        `function(){
            return resources.copper >= 50;
        }`,
        `function(){
            var ret = resources.copper >= 100;
            obj.mech[2] = (ret) ? obj.mech[2].map(x => x * 2) : obj.mech[2];
            resources.copper -= (ret) ? 100 : 0;
            return ret;
        }`,
    ]
]);